Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/TypeResolution/InteractableTypeFinder.cs

```diff
@@ -0,0 +1,140 @@
+        private static Dictionary<Type, List<InteractableType>> cache = new Dictionary<Type, List<InteractableType>>();
+
+        /// <summary>
+        /// Gets the list of InteractableType objects for classes that support the specified types.
+        /// </summary>
+        private static List<InteractableType> FindTypes(List<Type> types, TypeRestriction typeRestriction)
+        {
+            EnsureCacheForTypes(types, typeRestriction);
+            return GetTypesFromCache(types);
+        }
+
+        /// <summary>
+        /// Gets the list of InteractableType objects for classes that support the specified types by
+        /// looking directly in the cache.
+        /// </summary>
+        /// <remarks>
+        /// Assumes it is called after EnsureCacheForTypes. Otherwise, this is dangerous to call.
+        /// </remarks>
+        private static List<InteractableType> GetTypesFromCache(List<Type> types)
+        {
+            List<InteractableType> interactableTypes = new List<InteractableType>();
+            foreach (Type type in types)
+            {
+                interactableTypes.AddRange(cache[type]);
+            }
+            return interactableTypes;
+        }
+
+        /// <summary>
+        /// Ensures a cache entry is setup for all types in the InteractableType enum.
+        /// </summary>
+        /// <remarks>
+        /// Note that this is not invoked at runtime and is assumed to be invoked from a single
+        /// threaded UI context, and is thus not locked.
+        /// </remarks>
+        private static void EnsureCacheForTypes(List<Type> types, TypeRestriction typeRestriction)
+        {
+            List<Type> cacheMisses = new List<Type>();
+            foreach (Type type in types)
+            {
+                if (!cache.ContainsKey(type))
+                {
+                    cacheMisses.Add(type);
+                }
+            }
+
+            var assemblies = AppDomain.CurrentDomain.GetAssemblies();
+            foreach (Type type in cacheMisses)
+            {
+                cache[type] = GetTypesFromAssemblies(type, typeRestriction, assemblies);
+            }
+        }
+
+        /// <summary>
+        /// Loads the classes that derive from the given type by looking through all of the assemblies.
+        /// </summary>
+        private static List<InteractableType> GetTypesFromAssemblies(Type type, TypeRestriction typeRestriction, Assembly[] assemblies)
+        {
+            List<InteractableType> interactableTypes = new List<InteractableType>();
+            foreach (Assembly assembly in assemblies)
+            {
+                foreach (Type assemblyType in assembly.GetTypes())
+                {
+                    TypeInfo info = assemblyType.GetTypeInfo();
+                    bool exactBaseMatch = typeRestriction == TypeRestriction.AllowBase && assemblyType.Equals(type);
+                    bool derivedMatch = info.BaseType != null && info.BaseType.Equals(type);
```

Comment:
There is info.IsSubclassOf(type) that you can use here, as it may not be immediate base class.

Reply:
sg, will do that instead. This was taken from the original code.